### PR TITLE
feat(protractor): expose pending $http and $timeout on a timeout

### PR DESCRIPTION
**Pull request by hankduan**

### CODE REVIEW

**Comment by sjelin** [comment]
LGTM, I'm not an angular expert though

File: lib/protractor.js [hunk 1]
```diff
@@ -301,9 +302,42 @@ Protractor.prototype.waitForAngular = function(opt_description) {
...
       if (timeout) {
-        throw 'Timed out waiting for Protractor to synchronize with ' +
+        var errMsg = 'Timed out waiting for Protractor to synchronize with ' +
             'the page after ' + timeout + '. Please see ' +
             'https://github.com/angular/protractor/blob/master/docs/faq.md';
+        var pendingTimeoutsPromise = self.executeScript_(
+          'return window.pendingTimeouts',
+          'Protractor.waitForAngular() - getting pending timeouts' + description
+        );
+        var pendingHttpsPromise = self.executeScript_(
```
Comment (juliemr) [inline lib/protractor.js]: I think it's worth pulling this into clientsidescripts

File: lib/protractor.js [hunk 2]
```diff
@@ -414,6 +448,48 @@ Protractor.prototype.addBaseMockModules_ = function() {
           if ($compileProvider.debugInfoEnabled) {
             $compileProvider.debugInfoEnabled(true);
           }
+        }]).
+        config(['$provide', function($provide) {
+          $provide.decorator('$timeout', ['$delegate', function($delegate) {
+            var timeout = $delegate;
+
+            var taskId = 0;
+            if (!window.pendingTimeouts) {
```
Comment (juliemr) [inline lib/protractor.js]: Let's make this more debuggey and less likely to collide, what about `window.NG_PENDING_TIMEOUTS`

File: lib/protractor.js [hunk 3]
```diff
@@ -414,6 +448,48 @@ Protractor.prototype.addBaseMockModules_ = function() {
           if ($compileProvider.debugInfoEnabled) {
             $compileProvider.debugInfoEnabled(true);
           }
+        }]).
+        config(['$provide', function($provide) {
+          $provide.decorator('$timeout', ['$delegate', function($delegate) {
+            var timeout = $delegate;
```
Comment (juliemr) [inline lib/protractor.js]: I'd prefer `$timeout = $delegate`

File: lib/protractor.js [hunk 4]
```diff
@@ -301,9 +302,42 @@ Protractor.prototype.waitForAngular = function(opt_description) {
...
+                for (var key in pendingTimeouts) {
+                  if (pendingTimeouts.hasOwnProperty(key)) {
+                    pendingTasks.push('- $timeout: ' + pendingTimeouts[key]);
+                  }
+                }
+                for (var key in pendingHttps) {
+                  pendingTasks.push('- $http: ' + pendingHttps[key].url);
+                }
+                if (pendingTasks.length) {
+                  errMsg += '. Note the following pending task(s):\n';
```
Comment (juliemr) [inline lib/protractor.js]: What about "The following tasks were pending:"

File: spec/errorTest/baseCase/slow_http_and_timeout_spec.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+describe('success spec', function() {
```
Comment (juliemr) [inline spec/errorTest/baseCase/slow_http_and_timeout_spec.js]: what about `describe('slow asynchronous events'`?

**Comment by juliemr** [comment]
A couple style comments. Could you also add some info to the body of the commit message, like "Now when a test times out while waiting for Angular to be stable, pending $timeout and $http tasks will be reported to the console"

**Comment by hankduan** [comment]
address comments. PTAL

**Comment by juliemr** [comment]
Looks good!

**Comment by hankduan** [comment]
merged with https://github.com/angular/protractor/commit/6c10378b9a4e7cae9ba491a63ae11f942c833100

**Pull request closed without merging**